enforcer/datapath/datapath.go

```diff
@@ -223,32 +226,43 @@ func NewWithDefaults(
 // Enforce implements the Enforce interface method and configures the data path for a new PU
 func (d *Datapath) Enforce(contextID string, puInfo *policy.PUInfo) error {
 
-	puContext, err := d.contextTracker.Get(contextID)
-	if err != nil {
-		// Call proxy enforce from here and not from trireme like for other calls
-		// We will call enforce every time on the enforce so no need to propagate this info out of enforcer package
-		zap.L().Debug("Called Proxy Enforce")
-
-		proxyerr := d.proxyhdl.Enforce(contextID, puInfo)
-		if proxyerr != nil {
-			zap.L().Error("Unable to Enforce", zap.Error(proxyerr))
-			return proxyerr
-		}
+	zap.L().Debug("Called Proxy Enforce")
 
-		createerr := d.doCreatePU(contextID, puInfo)
-		if createerr != nil {
-			zap.L().Error("Called Do Create Returned error", zap.Error(createerr))
-			return createerr
-		}
+	// setup proxy before creating PU
+	if proxyerr := d.proxyhdl.Enforce(contextID, puInfo); proxyerr != nil {
+		return fmt.Errorf("Unable to Enforce- %s", proxyerr)
+	}
 
-		return nil
+	// Always create a new PU context
+	pu, err := pucontext.NewPU(contextID, puInfo, d.ExternalIPCacheTimeout)
+	if err != nil {
+		return errors.New("Error creating New PU")
 	}
-	//In case of update
-	proxyerr := d.proxyhdl.Enforce(contextID, puInfo)
-	if proxyerr != nil {
-		return proxyerr
+
+	// Cache PUs for retrieval based on packet information
+	if pu.Type() == constants.LinuxProcessPU || pu.Type() == constants.UIDLoginPU {
+		mark, ports := pu.GetProcessKeys()
+
+		d.puFromMark.AddOrUpdate(mark, pu)
+
+		for _, port := range ports {
+			d.puFromPort.AddOrUpdate(port, contextID)
+		}
+	} else {
+		if ip, ok := puInfo.Runtime.DefaultIPAddress(); ok {
+			d.puFromIP.AddOrUpdate(ip, pu)
+		} else {
+			d.puFromIP.AddOrUpdate(enforcerconstants.DefaultNetwork, pu)
+		}
 	}
-	return d.doUpdatePU(puContext.(*pucontext.PUContext), puInfo)
+
+	// Cache PU from contextID for management and policy updates
+	d.contextTracker.AddOrUpdate(contextID, pu)
+
+	// cache contextID from PU for reverse lookups.
+	d.contextIDTracker.AddOrUpdate(pu, contextID)
```

Comment:
Based on my comment above we use pu pointer to contextID - can we check if contextID is a part of pu and if not add it there - do we really need this map ?